### PR TITLE
publish-artifacts: Add GCS support

### PR DESCRIPTION
**Pull request by rail**
Previously, the only supported storage provider was S3.

This PR adds support for GCS.
* Make the bucket name be provider specific.
* The GCS bucket is optional for now.
* Move logic to separate files.
* No direct s3 imports from publish-artifacts and publish-provisional-artifacts.
* Create an extra abstraction interface to hide the s3-specific structs and calls.
* The "latest" URLs in publish-provisional-artifacts use S3's website redirect instead of copying the whole binary. This unifies it with publish-artifacts.

Release note: None

### CODE REVIEW

File: pkg/cmd/publish-provisional-artifacts/main.go [hunk 1]
```diff
@@ -36,47 +32,49 @@ const (
...
 	}
-	s3, err := makeS3()
+	var providers []release.ObjectPutGetter
+	s3, err := release.NewS3("us-east-1", bucketName)
 	if err != nil {
 		log.Fatalf("Creating AWS S3 session: %s", err)
 	}
+	providers = append(providers, s3)
+	if *gcsBucket != "" {
+		if _, ok := os.LookupEnv("GOOGLE_APPLICATION_CREDENTIALS"); !ok {
```
Comment (rickystewart) [inline pkg/cmd/publish-provisional-artifacts/main.go:68]: Nit: this logic is duplicated. You can refactor this out and just put the validation in `release.NewGCS()`.

**Review by rickystewart — APPROVED** [review]

File: pkg/cmd/publish-provisional-artifacts/main.go (hunk 1 above)
Comment (rail) [reply pkg/cmd/publish-provisional-artifacts/main.go:68]: Good point. Let me think if I want the function return an error or has the same signature as the S3 counterpart.

**Review by jlinder — APPROVED** [review]

**Comment by rail** [comment]
I tested this PR via a custom build and everything seems working. I'm going to merge this without adding GCS buckets in order to bake it for a few days. If everything goes as expected, I'm going to create another PR to enable GCS uploads. After some baking period we can backport these changes to the active release branches.

**Comment by rail** [comment]
bors r=rickystewart,jlinder

File: pkg/cmd/publish-provisional-artifacts/main.go (hunk 1 above)
Comment (rail) [reply pkg/cmd/publish-provisional-artifacts/main.go:68]: I left it as is to keep the "panic" logic out of the "library", similarly to the S3 approach.

**Pull request merged**